<!DOCTYPE html>
<html>
    <head>
        <meta charset="UTF-8">
        <title>ATAC-seq QC</title>
        <link rel="stylesheet" href="css/font-awesome.min.css">
        <link rel="stylesheet" href="css/normalize.css">
        <link rel="stylesheet" href="css/ataqc.css">
        <script src="js/dygraph-combined.js" defer></script>
        <script src="js/list.min.js" defer></script>
        <script src="js/ataqc.js" defer></script>
        <script src="js/data.js" defer></script>
    </head>
    <body>
        <div id="banner">
            <h1><a href="https://github.com/ParkerLab/ataqc/">ataqc</a></h1>
            <ul id="tabs" class="tabs">
                <li class="tab active" data-tab="experimentTabBody">
                    <h1>
                        <a class="tabLink" href="#experimentTabBody">Experiments
                            <svg class="badge" viewBox="0 0 8 8" version="1.1" xmlns="http://www.w3.org/2000/svg">
                                <defs>
                                    <radialGradient id="RadialGradient">
                                        <stop offset="0%" stop-color="#0cf"/>
                                        <stop offset="100%" stop-color="#037"/>
                                    </radialGradient>
                                </defs>
                                <circle cx="4" cy="4" r="4" fill="url(#RadialGradient)"/>
                            </svg>
                        </a>
                    </h1>
                </li>
                <li class="tab" data-tab="tableTabBody">
                    <h1>
                        <a class="tabLink" id="table_tab" href="#tableTabBody">Tables
                            <svg class="badge" viewBox="0 0 8 8" version="1.1" xmlns="http://www.w3.org/2000/svg">
                                <circle cx="4" cy="4" r="4" fill="url(#RadialGradient)"/>
                            </svg>
                        </a>
                    </h1>
                </li>
                <li class="tab" data-tab="plotTabBody">
                    <h1>
                        <a class="tabLink" id="plotTab" href="#plotTabBody">Plots
                            <svg class="badge" viewBox="0 0 8 8" version="1.1" xmlns="http://www.w3.org/2000/svg">
                                <circle cx="4" cy="4" r="4" fill="url(#RadialGradient)"/>
                            </svg>
                        </a>
                    </h1>
                </li>
                <li><h1><a id="help" href="#experimentTabBodyHelp" data-helpid="experimentTabBodyHelp">Help</a></h1></li>
            </ul>
        </div>
        <div id="main">
            <ul id="active_experiments">
            </ul>
            <div class="tabbodies">
                <div id="experimentTabBody" class="tabbody active">
                    <div id="experiments">
                        <div class="controls">
                            <button id="addAllExperiments"><span>Add</span></button>
                            <button id="removeAllExperiments"><span>Remove</span></button>
                            <div class="inputbox">
                                <input class="search" placeholder="Search" />
                                <button class="cleaner"><i class="fa fa-times-circle"></i><span></span></button>
                            </div>
                        </div>
                        <table>
                            <thead>
                                <tr>
                                    <th>
                                        Name
                                    </th>
                                    <th>
                                        <button class="sort" data-sort="organism"><span>Organism</span></button>
                                    </th>
                                    <th>
                                        Library description
                                    </th>
                                    <th>
                                        Experiment description
                                    </th>
                                    <th>
                                        ataqc metrics
                                    </th>
                                </tr>
                            </thead>
                            <tbody class="list">
                            </tbody>
                        </table>
                        <div class="footer">
                            Experiments measured with ataqc, from the Parker lab at the University of Michigan
                        </div>
                    </div>
                </div>
                <div id="tableTabBody" class="tabbody">
                    <div id="tables">
                        <div class="tableContainer">
                            <h2>Read Mapping Metrics</h2>
                            <table id="summary" class="data">
                                <thead>
                                    <tr>
                                        <th data-metric="name" class="name">
                                            <button class="sort" data-sort="name">
                                                Experiment
                                            </button>
                                        </th>
                                        <th data-metric="total_reads" data-metrictype="integer" data-hovertext="The total number of reads from the experiment.">
                                            <button class="sort" data-sort="total_reads"><span>Total reads</span></button>
                                        </th>
                                        <th data-metric="properly_paired_and_mapped_reads" data-percentage="total_reads" data-hovertext="Properly paired and mapped reads, as a percentage of all reads.">
                                            <button class="sort" data-sort="properly_paired_and_mapped_reads"><span>Properly paired and mapped&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="secondary_reads" data-percentage="total_reads" data-hovertext="Alternate alignments of reads, of lower quality than their primary alignments.">
                                            <button class="sort" data-sort="secondary_reads"><span>Secondary&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="supplementary_reads" data-percentage="total_reads" data-hovertext="Aligned segments of reads mapped in chimeric alignments.">
                                            <button class="sort" data-sort="supplementary_reads" ><span>Supplementary&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="duplicate_reads" data-percentage="total_reads" data-hovertext="Duplicate alignments as a percentage of total reads.">
                                            <button class="sort" data-sort="duplicate_reads"><span>Duplicate&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="mean_mapq" data-hovertext="The mean mapping quality of the experiment's reads.">
                                            <button class="sort" data-sort="mean_mapq"><span>Mean mapping quality</span></button>
                                        </th>
                                        <th data-metric="median_mapq" data-metrictype="integer" data-hovertext="The median mapping quality of the experiment's reads.">
                                            <button class="sort" data-sort="median_mapq"><span>Median mapping quality</span></button>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody class="list">
                                </tbody>
                            </table>
                        </div>

                        <div class="tableContainer">
                            <h2>Quality Indicators</h2>
                            <table id="indicators" class="data">
                                <thead>
                                    <tr>
                                        <th data-metric="name" class="name">
                                            <button class="sort" data-sort="name">
                                                Experiment
                                            </button>
                                        </th>
                                        <th data-metric="hqaa" data-percentage="total_reads"  data-hovertext="High quality autosomal alignments, as a percentage of all reads. We define these as the primary (not marked duplicate, supplementary, or secondary) alignments of reads properly paired and mapped to autosomes with mapping quality of at least 30. If this metric seems low, look for mitochondrial contamination in the reference metrics below.">
                                            <button class="sort" data-sort="hqaa"><span>High quality autosomal alignments&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="short_mononucleosomal_ratio" data-hovertext="The ratio of fragments 50-100bp long to fragments 150-200bp long. This can indicate under- or over-transposition -- which should also be apparent on the fragment length distribution plot.">
                                            <button class="sort" data-sort="short_mononucleosomal_ratio"><span>Short-to-mononucleosomal ratio</span></button>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody class="list">
                                </tbody>
                            </table>
                        </div>

                        <div class="tableContainer">
                            <h2>Unmapped Read Metrics</h2>
                            <table id="basic_problems" class="data">
                                <thead>
                                    <tr>
                                        <th data-metric="name" class="name">
                                            <button class="sort" data-sort="name"><span>Experiment</span></button>
                                        </th>
                                        <th data-metric="unmapped_reads" data-percentage="total_reads" data-hovertext="Unmapped reads, as a percentage of all reads.">
                                            <button class="sort" data-sort="unmapped_reads"><span>Unmapped&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="unmapped_mate_reads" data-percentage="total_reads" data-hovertext="Unmapped mate reads, as a percentage of all reads.">
                                            <button class="sort" data-sort="unmapped_mate_reads"><span>Unmapped mate&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="qcfailed_reads" data-percentage="total_reads" data-hovertext="Reads failing quality controls, as a percentage of all reads.">
                                            <button class="sort" data-sort="qcfailed_reads"><span>QC&nbsp;fail&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="unpaired_reads" data-percentage="total_reads" data-hovertext="Unpaired reads, as a percentage of all reads.">
                                            <button class="sort" data-sort="unpaired_reads"><span>Unpaired&nbsp;(%)</span></button>
                                        </th>
                                        <th data-metric="reads_mapped_with_zero_quality" data-percentage="total_reads" data-hovertext="Alignments with mapping quality of zero, as a percentage of all reads.">
                                            <button class="sort" data-sort="reads_mapped_with_zero_quality"><span>Zero mapping quality&nbsp;(%)</span></button>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody class="list">
                                </tbody>
                            </table>
                        </div>

                        <div class="tableContainer">
                            <h2>Aberrant Mapping Metrics</h2>
                            <table id="other_problems" class="data">
                                <thead>
                                    <tr>
                                        <th data-metric="name" class="name" rowspan="2">
                                            <button class="sort" data-sort="name"><span>Experiment</span></button>
                                        </th>
                                        <th colspan="7">Of all reads, the percentage that paired and mapped but:</th>
                                    </tr>

                                    <tr>
                                        <th data-metric="rf_reads" data-percentage="total_reads" data-hovertext="The reads of a proper pair should point toward each other, with the end mapped to the smaller coordinate on the forward strand and the other end on the reverse strand -- FR orientation. These reads were in pairs oriented RF, where the smaller coordinate was on the reverse strand and the larger on the forward strand.">
                                            <button class="sort" data-sort="rf_reads"><span>RF</span></button>
                                        </th>
                                        <th data-metric="ff_reads" data-percentage="total_reads" data-hovertext="The reads of a proper pair should point toward each other, with the end mapped to the smaller coordinate on the forward strand and the other end on the reverse strand -- FR orientation. These reads were in pairs oriented FF, where both reads mapped on the forward strand.">
                                            <button class="sort" data-sort="ff_reads"><span>FF</span></button>
                                        </th>
                                        <th data-metric="rr_reads" data-percentage="total_reads" data-hovertext="The reads of a proper pair should point toward each other, with the end mapped to the smaller coordinate on the forward strand and the other end on the reverse strand -- FR orientation. These reads were in pairs oriented RR, where both reads mapped on the reverse strand.">
                                            <button class="sort" data-sort="rr_reads"><span>RR</span></button>
                                        </th>
                                        <th data-metric="reads_with_mate_mapped_to_different_reference" data-percentage="total_reads" data-hovertext="Each read of the pair mapped to a different reference (chromosome).">
                                            <button class="sort" data-sort="reads_with_mate_mapped_to_different_reference"><span>on different chromosomes</span></button>
                                        </th>
                                        <th data-metric="reads_with_mate_too_distant" data-percentage="total_reads" data-hovertext="The reads of the pair aligned to locations that are too far apart to be plausible, given the fragment lengths in the dataset.">
                                            <button class="sort" data-sort="reads_with_mate_too_distant"><span>too far from mate</span></button>
                                        </th>
                                        <th data-metric="reads_mapped_and_paired_but_improperly" data-percentage="total_reads" data-hovertext="We don't know why these are not marked properly paired and mapped. Sorry. You can have ataqc log problematic reads, so you can try to diagnose them. Let us know what you find.">
                                            <button class="sort" data-sort="reads_mapped_and_paired_but_improperly"><span>just not properly</span></button>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody class="list">
                                </tbody>
                            </table>
                        </div>

                        <div class="tableContainer">
                            <h2>Reference Mapping Metrics</h2>
                            <table id="reference" class="data">
                                <thead>
                                    <tr>
                                        <th rowspan="2" data-metric="name" class="name">
                                            <button class="sort" data-sort="name">Experiment</button>
                                        </th>
                                        <th colspan="2">Autosomal</th>
                                        <th colspan="2">Mitochondrial</th>
                                    </tr>

                                    <tr>
                                        <th data-metric="total_autosomal_reads" data-percentage="total_reads" class="rj" data-hovertext="The number of reads aligned to autosomes, as a percentage of all reads.">
                                            <button class="sort" data-sort="total_autosomal_reads"><span>% of all reads</span></button>
                                        </th>
                                        <th data-metric="duplicate_autosomal_reads" data-percentage="total_autosomal_reads" data-hovertext="The percentage of reads mapped to autosomes that were duplicates.">
                                            <button class="sort" data-sort="duplicate_autosomal_reads"><span>Duplicates (% of autosomal alignments)</span></button>
                                        </th>
                                        <th data-metric="total_mitochondrial_reads" data-percentage="total_reads" class="rj" data-hovertext="The number of reads aligned to the mitochondrial reference, as a percentage of all reads.">
                                            <button class="sort" data-sort="total_mitochondrial_reads"><span>% of all reads</span></button>
                                        </th>
                                        <th data-metric="duplicate_mitochondrial_reads" data-percentage="total_mitochondrial_reads" data-hovertext="The percentage of reads mapped to the mitochondrial reference that were duplicates.">
                                            <button class="sort" data-sort="duplicate_mitochondrial_reads"><span>Duplicates&nbsp;(% of mitochondrial alignments)</span></button>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody class="list">
                                </tbody>
                            </table>
                        </div>

                        <div class="tableContainer">
                            <h2>Peak Metrics</h2>
                            <table id="peaks" class="data">
                                <thead>
                                    <tr>
                                        <th data-metric="name" class="name">
                                            <button class="sort" data-sort="name">Experiment</button>
                                        </th>
                                        <th data-metric="total_peaks" data-metrictype="integer" data-hovertext="The total number of peaks called.">
                                            <button class="sort" data-sort="total_peaks">Total peaks</button>
                                        </th>
                                        <th data-metric="hqaa_overlapping_peaks_percent" data-hovertext="The percentage of high quality autosomal alignments that overlapped called peaks.">
                                            <button class="sort" data-sort="hqaa_overlapping_peaks_percent">High quality autosomal alignments overlapping peaks&nbsp;(%)</button>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody class="list">
                                </tbody>
                            </table>
                        </div>
                        <div class="footer">
                            Metrics produced with ataqc, from the Parker lab at the University of Michigan
                        </div>
                    </div>
                </div>
                <div id="plotTabBody" class="tabbody">
                    <div id="plots">
                        <div class="plotContainer" id="fragmentLength">
                            <h2>Fragment length distribution</h2>
                            <div class="plotbox">
                                <div class="plot"></div>
                                <div class="legend"></div>
                            </div>
                        </div>
                        <div class="plotContainer" id="mapq">
                            <h2>Mapping quality distribution</h2>
                            <div class="plotbox">
                                <div class="plot"></div>
                                <div class="legend"></div>
                            </div>
                        </div>
                        <div class="plotContainer" id="peakReadCounts">
                            <h2>Cumulative fraction of high quality autosomal alignments</h2>
                            <div class="plotbox">
                                <div class="plot"></div>
                                <div class="legend"></div>
                            </div>
                        </div>
                        <div class="plotContainer" id="peakTerritory">
                            <h2>Cumulative fraction of peak territory</h2>
                            <div class="plotbox">
                                <div class="plot"></div>
                                <div class="legend"></div>
                            </div>
                        </div>
                    </div>
                    <div class="footer">
                        Plots made with ataqc, from the Parker lab at the University of Michigan
                    </div>
                </div>
            </div>
            <div class="help" id="experimentTabBodyHelp">
                <h2><div>Using the Experiments tab</div><button class="cleaner"><i class="fa fa-times-circle"></i><span></span></button></h2>
                <p>Click the experiment name in the first column to add it to the list of active experiments shown in the Tables and Plots tabs.</p>
                <p>If a URL was specified when running ataqc, the "Experiment description" column will contain links to more information on the experiments.</p>
                <p>The original ataqc metrics files used to produce this viewer can be downloaded via the links in the last column. They're plain JSON files that you can use to produce your own plots with R or Pandas, or analyze further with your own programs.</p>
                <p>You can search the experiment list by entering text in the Search box. Clear the search with the <i class="fa fa-times-circle"></i> icon.</p>
                <p>Click the <button>Add</button> button to add all displayed experiments to the active experiment list, so you can compare them in the Tables and Plots tabs. If you've entered text in the Search box, only the matching experiments will be addedl.</p>
                <p>If there are any active experiments, the <button>Remove</button> button will remove the entire set, or if you've entered text in the Search box, just the ones that match.</p>
            </div>
            <div class="help" id="tableTabBodyHelp">
                <h2><div>Using the Tables tab</div><button class="cleaner"><i class="fa fa-times-circle"></i><span></span></button></h2>
                <p>Hover over table headings to see descriptions of the metrics.</p>
                <p>Most table headings can be clicked to sort the table.</p>
                <p>Hovering over any table row will highlight it in all
                    the tables. This works best, of course, if you're not
                    comparing too many experiments to fit on the screen.</p>
                <p>The maximum value in any column is highlighted in gold text.</p>
            </div>
            <div class="help" id="plotTabBodyHelp">
                <h2><div>Using the Plots tab</div><button class="cleaner"><i class="fa fa-times-circle"></i><span></span></button></h2>
                <p>Hover over plot lines or their legend entries to highlight them.</p>
                <p>When hovering over a plot, its legend is sorted by the Y values at that X position.</p>
                <p>Click and drag within a plot to zoom in on a region.</p>
                <p>Double-click the plot at any time to reset it.</p>
            </div>
        </div>
        <script>
            function initialize() {
                ataqc.initialize('experiments');
            };

            function ready(fn) {
                if (document.readyState != 'loading') {
                    fn();
                } else {
                    document.addEventListener('DOMContentLoaded', fn);
                }}
            ready(initialize);
        </script>
    </body>
</html>
